Public/Src/Engine/UnitTests/Scheduler/FileMonitoringViolationAnalyzerTests.cs

```diff
@@ -636,7 +637,91 @@ public void AggregateLogMessageOnUndeclaredReadWrite()
+        {
+            var testContext = BuildXLContext.CreateInstanceForTesting();
+            var tool = AbsolutePath.Create(testContext.PathTable, X("/X/out/tool.exe"));
+            var path = AbsolutePath.Create(testContext.PathTable, X("/X/out/path.h"));
+
+            HashSet<ReportedViolation> violations = new HashSet<ReportedViolation>();
+            foreach (DependencyViolationType violationType in violationTypes)
+            {
+                violations.Add(new ReportedViolation(isError: false, type: violationType, path: path, violatorPipId: new PipId(1), PipId.Invalid, tool));
+            }
+
+            string result = FileMonitoringViolationAnalyzer.AggregateAccessViolationPaths(violations, testContext.PathTable, (pipId) => "dummy");
+            string accessLine = result.Split(new string [] { Environment.NewLine}, StringSplitOptions.None)[1].TrimStart(' ');
+
+            XAssert.IsTrue(accessLine.StartsWith(expectedPrefix), "Unexpected result: " + result);
+        }
+
+        [Fact]
+        public void AnalyzerMessageRenderingLegendInformation()
+        {
+            var testContext = BuildXLContext.CreateInstanceForTesting();
+            var tool1 = AbsolutePath.Create(testContext.PathTable, X("/X/out/tool1.exe"));
+            var tool2 = AbsolutePath.Create(testContext.PathTable, X("/X/out/tool2.exe"));
+            var pathA = AbsolutePath.Create(testContext.PathTable, X("/X/out/pathA.h"));
+            var pathB = AbsolutePath.Create(testContext.PathTable, X("/X/out/pathB.h"));
+
+            HashSet<ReportedViolation> violations = new HashSet<ReportedViolation>();
+            violations.Add(new ReportedViolation(isError: false, type: DependencyViolationType.DoubleWrite, path: pathA,
+                violatorPipId: new PipId(1), relatedPipId: PipId.Invalid, tool1));
+            violations.Add(new ReportedViolation(isError: false, type: DependencyViolationType.AbsentPathProbeUnderUndeclaredOpaque, path: pathB,
+                violatorPipId: new PipId(1), relatedPipId: new PipId(2), tool1));
+            violations.Add(new ReportedViolation(isError: false, type: DependencyViolationType.WriteInExistingFile, path: pathB,
+                violatorPipId: new PipId(1), relatedPipId: PipId.Invalid, tool2));
+            violations.Add(new ReportedViolation(isError: false, type: DependencyViolationType.WriteInExistingFile, path: pathB,
+                violatorPipId: new PipId(1), relatedPipId: PipId.Invalid, tool2));
+
+            string result = FileMonitoringViolationAnalyzer.AggregateAccessViolationPaths(violations, testContext.PathTable, (pipId) => "PLACEHOLDER PIP DESCRIPTION");
+
+            // Should see a single instance of " W " for the write files legend even though there are 2 write violations
+            XAssert.AreEqual(1, CountInstancesOfWord(" W ", result), result);
+
+            XAssert.AreEqual(1, CountInstancesOfWord(" DW ", result), result);

```

Comment:
Nit: Maybe extract those to some readonly strings and expose internally so we dont need to fix up unit tests when we change the legend text?